### PR TITLE
Losses

### PR DESCRIPTION
**Pull request by robertodessi**
Adding code for 3 standard losses: reconstruction, discriminative, ntxent (contrastive).

## Description
Adding 3 standard losses as class objects.

## Motivation and Context
Common losses in emergent communication games are contrastive and discriminative loss.
Most of the games in EGG used either of the two. Adding them in core so that it will be easier to implement a new game from scratch. No more need to implement a discriminative loss (a potential source of wasted time handling tensor dimensions between sender messages and processed receiver input) as it can be imported from core in one line.

## TODOs (potential)
rewrite some games with such standard losses

### CODE REVIEW

File: egg/core/losses.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+
+class RecoLoss:
+    def __init__(self, n_attributes: int, n_values: int, batch_size: int):
+        self.n_attributes = n_attributes
+        self.n_values = n_values
+        self.batch_size = batch_size
+
+    def __call__(self, sender_input, _message, _receiver_input, receiver_output, labels):
+        self.reconstruction_loss(
```
Comment (eugene-kharitonov) [inline egg/core/losses.py:29]:
```suggestion
        return self.reconstruction_loss(
```

File: egg/core/losses.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        return self.discrimination_loss(receiver_output, labels)
+
+    @staticmethod
+    def discrimination_loss(receiver_output, labels):
+        acc = (receiver_output.argmax(dim=1) == labels).detach().float()
+        loss = F.cross_entropy(receiver_output, labels, reduction="none")
+        return loss, {'acc': acc}
+
+
+class RecoLoss:
```
Comment (eugene-kharitonov) [inline egg/core/losses.py:22]: Let's unroll the name? I was thinking it is some _recommendation_ loss :)

File: egg/core/losses.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        loss = loss.view(batch_size, -1).mean(dim=1)
+        return loss, {'acc': acc}
+
+
+class NTXentLoss:
+    def __init__(
+            self,
+            batch_size: int,
+            temperature: float = 0.1,
+            device: torch.device = torch.device("cuda" if torch.cuda.is_available() else "cpu"),
```
Comment (eugene-kharitonov) [inline egg/core/losses.py:60]: What do you think if we wouldn't store it as a part of the loss function instance, but decide on the fly when running forward()? Also I think it is not used now.

File: egg/core/losses.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+            similarity: str = "cosine",
+            normalize_before: bool = True
+    ):
+        self.temperature = temperature
+        self.batch_size = batch_size
+        self.device = device
+
+        similarities = {"cosine", "dot"}
+        assert similarity.lower() in similarities, "Cannot recognize similarity function {similarity}"
```
Comment (eugene-kharitonov) [inline egg/core/losses.py:69]:
```suggestion
        assert similarity.lower() in similarities, f"Cannot recognize similarity function {similarity}"
```

File: egg/core/losses.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        receiver_guesses = receiver_output.argmax(dim=1)
+        correct_samples = (receiver_guesses == labels.view(-1)).view(batch_size, n_attributes).detach()
+        acc = (torch.sum(correct_samples, dim=-1) == n_attributes).float()
+        labels = labels.view(batch_size * n_attributes)
+        loss = F.cross_entropy(receiver_output, labels, reduction="none")
+        loss = loss.view(batch_size, -1).mean(dim=1)
+        return loss, {'acc': acc}
+
+
+class NTXentLoss:
```
Comment (eugene-kharitonov) [inline egg/core/losses.py:55]: Unlike other losses, this thing is huuge. Do you think it will make sense to add a short description/py doc?

File: egg/core/losses.py (hunk 3 above)
Comment (robertodessi) [reply egg/core/losses.py:60]: It's used on line 102 but you're right in that is should be computed on the fly. I will update it accordingly

File: egg/core/losses.py (hunk 5 above)
Comment (robertodessi) [reply egg/core/losses.py:55]: yep, I would like to add some doctest as well, especially if we rewrite some games using these

File: egg/core/losses.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    def __init__(
+            self,
+            batch_size: int,
+            temperature: float = 0.1,
```
Comment (eugene-kharitonov) [inline egg/core/losses.py:59]: [nit] (it seems i've lost this comment?) 0.1 is pretty unusual default temp, I'd think we should either have 1.0 or not have a default at all. Generally, I think defaults should be very low-surprise and for temperatures, in my eyes, low-surprise is 1.0...

File: egg/core/losses.py (hunk 3 above)
Comment (eugene-kharitonov) [reply egg/core/losses.py:60]: L102: `positive_samples.device`?

Comment (robertodessi) [reply egg/core/losses.py:60]: uh, never mind you're right it is not used!

File: egg/core/losses.py (hunk 6 above)
Comment (robertodessi) [reply egg/core/losses.py:59]: Based on the simclr paper it seems 1.0 is used but not really as a default value, if I remember correcty it gave better contrastive accuracy but not linear eval accuracy.  Would you still consider 1.0 low-surprise for this specific loss?

**Comment by robertodessi** [comment]
Thanks a lot for the review! @eugene-kharitonov
I'll merge once I can write some proper doc and testing
